### PR TITLE
feature(avatar): avatar is now served by the new API

### PR DESCRIPTION
**Pull request by hypeJunction**
Avatars are now served using elgg_get_inline_url()
- [x] Requires #9147

### CODE REVIEW

**Comment by mrclay** [comment]
Looks good

**Comment by hypeJunction** [comment]
This doesn't need a rebase. Just tested again manually, avatars are served as expected.

File: mod/profile/icondirect.php [hunk 1]
```diff
@@ -6,6 +6,8 @@
  * @package ElggProfile
  */
 
+elgg_deprecated_notice("icondirect.php is no longer used and will be removed, do not include and require it. Use elgg_get_inline_url() instead.", '2.1');
+
```
Comment (mrclay) [inline mod/profile/icondirect.php]: Should this just create the new image URL and forward to it?

Comment (hypeJunction) [inline mod/profile/icondirect.php]: Yes, that's what I did with groups. Will do that.

**Comment by mrclay** [comment]
What if a plugin has changed the default user icons via `entity:icon:url` or the `resource/view/avatar` view? Won't this break that?

**Comment by hypeJunction** [comment]
I don't see how. Their views/handlers will take precedence.

**Comment by mrclay** [comment]
OK, yes I'm seeing that.

File: engine/lib/users.php [hunk 2]
```diff
@@ -449,10 +449,16 @@ function user_avatar_hook($hook, $entity_type, $returnvalue, $params) {
 	$size = $params['size'];
 
 	if (isset($user->icontime)) {
-		return "avatar/view/$user->username/$size/$user->icontime";
```
Comment (mrclay) [inline engine/lib/users.php]: If a plugin *only* overrides the view, their view will never be called.

Comment (mrclay) [inline engine/lib/users.php]: Damn it, I was seeing this as a view render, not a URL. Sorry, ignore me.

Comment (mrclay) [inline engine/lib/users.php]: No, that's right. If the default changes to not using the resource view, that's a problem for BC.

Comment (mrclay) [inline engine/lib/users.php]: Ideally we could detect whether `resource/avatar/view` has been overridden and use it in that case.

Comment (hypeJunction) [inline engine/lib/users.php]: Hmm, let me dig into it

Comment (hypeJunction) [inline engine/lib/users.php]: We can leave the URL alone and forward from the resource view. Not sure how browsers will respond to caching with 301 headers...

Comment (mrclay) [inline engine/lib/users.php]: In browsers where it didn't work it would be pretty bad. Seems risky.

Comment (mrclay) [inline engine/lib/users.php]: This doesn't seem too tough: https://github.com/Elgg/Elgg/issues/9469#issuecomment-194861328

Comment (hypeJunction) [inline engine/lib/users.php]: This particular view is not really extendable, it sends headers so if you output content before or after, you would be making a mess. We could just check view location. Trying that now.

Comment (hypeJunction) [inline engine/lib/users.php]: Never mind, I see that we have to take into account the 'view' and 'view_vars' hooks...

**Comment by hypeJunction** [comment]
icondirect now forwards to the new url

**Comment by hypeJunction** [comment]
@mrclay Updated to use `_elgg_view_may_be_altered()`

File: engine/lib/users.php [hunk 3]
```diff
@@ -445,14 +445,28 @@ function user_create_hook_add_site_relationship($event, $object_type, $object) {
...
+		return;
+	}
+
+	$default_url = elgg_get_simplecache_url("icons/user/default{$size}.gif");
+	if (!isset($user->icontime)) {
+		return $default_url;
+	}
+
+	if (_elgg_view_may_be_altered('resources/avatar/view', 'resources/avatar/view.php')) {
 		return "avatar/view/$user->username/$size/$user->icontime";
```
Comment (mrclay) [inline engine/lib/users.php]: We need a little comment "For BC with 2.0 if a plugin is suspected of using this view/page handler we need to use it."

Comment (hypeJunction) [inline engine/lib/users.php]: Will add. FYI, tested manually by adding a resource view in a plugin. Didn't test with hooks.

Comment (hypeJunction) [inline engine/lib/users.php]: Done

**Comment by hypeJunction** [comment]
Added a deprecation commit.

File: docs/admin/upgrading.rst [hunk 4]
```diff
@@ -43,6 +43,12 @@ Deprecations in 2.x
 2.2
 ---
 
+User avatars are now served via ``serve-file`` handler. Plugins should start using ``elgg_get_inline_url()`` and note that:
+
+ * `/avatar/view` page handler and resource view have been deprecated
```
Comment (mrclay) [inline docs/admin/upgrading.rst]: Double backticks

Comment (mrclay) [inline docs/admin/upgrading.rst]: Besides that LGTM

**Comment by hypeJunction** [comment]
@mrclay updated docs to use double backticks

**Pull request merged**